WIP: Use hf datasets for activation store

tom-pollak opened the pull request:
Related issue: #320 

## Motivation

Use huggingface Dataset rather than safetensors caching activations ActivationStore.

- Consistent format, easily to load and share on hf hub.
- Caching from multiple hook points at once, with different shapes, in a single dataset (as different columns)
- Simplified: Sharding and memory mapping is default in `datasets.load_from_disk
- Datasets can be loaded as IterableDataset / streaming=True / n_shards, much simpler than doing manually.
- Shuffling support would be [easier](https://huggingface.co/docs/datasets/v3.0.1/en/package_reference/main_classes#datasets.IterableDataset.shuffle)

---

Code is written so that adding extra hook points is trivial.

## `CacheActivationsRunner`
For each hook point (only 1 available so far)  `Array2D(shape=(self.cfg.context_size, self.cfg.d_in), dtype=self.cfg.dtype`

Create single shard from `buffer = store.get_buffer()` with `_create_shard()`. This saves to `shards_{i}`

### `run()`

Incremenatally creates shards

When dataset is complete it concatenates shards (mem map), shuffles (in huggingface this optimizes to changing the indices without changing the underlying data)

If push to hf we resave the dataset which puts it in a nicer format (looong operation). Deletes the original temp shards.

- Possible performance improvements from using `IterableDataset` rather than standard.

### Misc

small fix: duplicate definitions of `tokens_in_buffer` in `__str__` and `run` -> moved to` __init__`

Set shuffle parameters in config to DEPRECATED = -1.

## Activation Store

This approach simplifies a lot, just load the hf dataset in `__init__` and have a `current_idx = 0` to determine the current idx in the dataset. Memory mapping & sharding is all handled automatically.

Activation dataset can be pushed to hub by passing 'hf_repo_id' in runner config.

I would also add the ability to load dataset from hf hub in ActivationStore, pretty simple change just by using `datasets.load_dataset` with a repo path rather than `datasets.load_from_disk` but this pr is big enough as is.

## tests

[Favourite test](https://github.com/jbloomAus/SAELens/pull/321/files#:~:text=test_compare_cached_activations_end_to_end_with_ground_truth)
- Runs activations caching pipeline without shuffling and `torch.allclose` the resulting hf dataset with manually taking the activations out of `model.run_with_cache`

Also added a bunch more tests to make sure it handles error cases correctly:

- Non-empty dataset dir
- Incorrect `d_in` when creating the dataset
- Incorrect `d_in` / `context_size` / `hook_name` when loading dataset

## benchmarking

[Benchmarking hf datasets vs safetensors](https://github.com/jbloomAus/SAELens/pull/321/files#diff-a535046d6dd14f343f3487cbdaa73b08bf92a3721c183802b49642c510dba740:~:text=test_hf_dataset_save_vs_safetensors)

# TODO: update docs / examples (will do post review)

## Type of change

Please delete options that are not relevant.

- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

Breaking change: no public api changes (except deprecating shuffle params) however this would break saved activations stores using the old method.

# Checklist:

- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I have not rewritten tests relating to key interfaces which would affect backward compatibility


### You have tested formatting, typing and unit tests (acceptance tests not currently in use)

- [X] I have run `make check-ci` to check format and linting. (you can run `make format` to format code if needed.)

jbloomAus commented:
I'm quite excited about this. Will try to review ASAP :)

tom-pollak commented:
Should be formatted correctly now :)

tom-pollak commented:
Ok done, no more changes

chanind reviewed (commented):
Really like this idea in general, sounds like there's a lot of advantages to moving to hf datasets. I'm a bit worried that there's no test which saves and then loads a dataset in the activations store, but I suppose the fixtures probably cover that to some extent.

Otherwise, just some minor comments on the PR but I'd be in favor of merging aside from those!

tom-pollak commented on sae_lens/config.py:
```diff
@@ -470,11 +472,16 @@ class CacheActivationsRunnerConfig:
+    shuffle_every_n_buffers: int = -1  # DEPRECATED
+    n_shuffles_with_last_section: int = -1  # DEPRECATED
+    n_shuffles_in_entire_dir: int = -1  # DEPRECATED
+    n_shuffles_final: int = -1  # DEPRECATED
```
👍

tom-pollak commented on sae_lens/cache_activations_runner.py:
```diff
@@ -73,94 +102,52 @@ def run(self):
+                revision=self.cfg.hf_revision,
+            )
+
+        return dataset
```
I'll add this, had this originally, but I didn't want to do a redundant `save_to_disk` if pushing to hub saves as a parquet dataset anyway

chanind replied:
That's fair; we could add a `save_to_disk` option to the runner config if we want to skip this and just upload. Still fine as is though IMO!

tom-pollak commented on sae_lens/cache_activations_runner.py:
```diff
@@ -73,94 +102,52 @@ def run(self):
+        print("Concatenating shards...")
+        dataset = concatenate_datasets(dataset_shards)
+        # for better performance:
+        # .to_iterable_dataset(num_shards=self.n_buffers)
```
I'll delete, can't save_to_disk / `push_to_hub` with `IterableDataset`

tom-pollak commented on sae_lens/training/activations_store.py:
```diff
@@ -475,9 +491,56 @@ def get_activations(self, batch_tokens: torch.Tensor):
+        assert self.cached_activation_dataset is not None
+        # In future, could be a list of multiple hook names
+        hook_names = [self.hook_name]
+        assert set(hook_names).issubset(self.cached_activation_dataset.column_names)
```
can remove, this check was also made when loading the dataset, added an assert just to be extra safe

tom-pollak commented on sae_lens/cache_activations_runner.py:
```diff
@@ -73,94 +102,52 @@ def run(self):
+
+        if self.cfg.hf_repo_id:
+            print("Pushing to hub...")
+            dataset.push_to_hub(
```
added!

tom-pollak commented:
replied to comments, added fixes. 

> Really like this idea in general, sounds like there's a lot of advantages to moving to hf datasets. I'm a bit worried that there's no test which saves and then loads a dataset in the activations store, but I suppose the fixtures probably cover that to some extent.

Deleted cached dataset in repo, it now generates it on the fly with `_create_dataset`

chanind commented on tests/unit/training/test_cache_activations_runner.py:
```diff
@@ -216,3 +215,239 @@ def W_E(self) -> torch.Tensor:
+    print(
+        "Max difference:",
+        torch.max(torch.abs(ground_truth_acts - activation_dataset[cfg.hook_name])),
+    )
```
It's not a huge deal, but would still prefer to remove these. If the test is flaky we should fix it so it's not flaky, but it looks like you've made sure it runs properly and that the bounds are reasonably tuned. In the same way that it's not ideal to have debugging prints left in production code, it just makes it feel like the code in the test isn't finished if it has debugging print statements in it, like leaving scaffolding attached to a building after construction has finished.

chanind reviewed (approved):
Nice work! This LGTM, I'll leave it to @jbloomAus to do the final merge if he's happy with this. We should make sure to do a major version bump as this is technically a breaking change.

tom-pollak commented:
Awesome thanks! CI seems to run out of space on testing it can load from file, can try to reduce the dataset size and retry.

tom-pollak commented:
@chanind @jbloomAus fixed out of space error when constructing test datasets (just made it smaller). Should be ready to go, can we re-run ci?
